File: internal/plugins/rstest/rules/no_disabled_tests/no_disabled_tests_test.go
package no_disabled_tests_test

import (
	"testing"

	"github.com/web-infra-dev/rslint/internal/plugins/rstest/fixtures"
	"github.com/web-infra-dev/rslint/internal/plugins/rstest/rules/no_disabled_tests"
	"github.com/web-infra-dev/rslint/internal/rule_tester"
)

func TestNoDisabledTestsRule(t *testing.T) {
	rule_tester.RunRuleTester(
		fixtures.GetRootDir(),
		"tsconfig.json",
		t,
		&no_disabled_tests.NoDisabledTestsRule,
		[]rule_tester.ValidTestCase{
			{Code: `test("case", () => {}); it("case", () => {}); describe("suite", () => {});`},
			{Code: `test.only("case", () => {}); test.fails("case", () => {}); test.concurrent.sequential("case", () => {});`},
			{Code: `describe.only("suite", () => {}); describe.concurrent.sequential("suite", () => {});`},
			{Code: `test.todo("case"); it.todo("case"); describe.todo("suite");`},
			{Code: `test.todo.each([1])("case"); test.todo.for([1])("case");`},
			{Code: `const todoTest = test.todo; todoTest("case");`},
			{Code: `test.skipIf(condition)("case", () => {}); test.runIf(condition)("case", () => {});`},
			{Code: `describe.skipIf(condition)("suite", () => {}); describe.runIf(condition)("suite", () => {});`},
			{Code: `test.skipIf(true)("case", () => {}); test.runIf(false)("case", () => {});`},
			{Code: `test("case", context => { context.skip(); });`},
			{Code: `test("case", ctx => { ctx.skip(); }); request.skip();`},
			{Code: `describe("empty suite"); describe.concurrent("empty suite");`},
			// an identifier is not necessarily an options object
			{Code: `test("case", options);`},
			{Code: `test("case", { timeout: 100 }, () => {});`},
			{Code: `test.todo("case", { timeout: 100 });`},
			{Code: `fit("case", () => {}); xit("case", () => {}); xtest("case", () => {}); fdescribe("suite", () => {}); xdescribe("suite", () => {});`},
			{Code: `pending();`},
			{Code: `import { pending } from "actions"; pending();`},
			{Code: `test[` + "`sk${suffix}`" + `]("case", () => {});`},
			{Code: `import { test } from "node:test"; test.skip("case", () => {});`},
			{Code: `import { test, describe } from "vitest"; test.skip("case", () => {}); describe.skip("suite", () => {});`},
			{Code: `const test = createRunner(); test.skip("case", () => {});`},
			{Code: `let skipped = test.skip; skipped("case", () => {});`},
			// a variable declaration without an initializer must not be treated as a require
			{Code: `declare const skipped: any; skipped("case", () => {});`},
			{Code: `declare const pending: any; pending();`},
			{Code: `import { customTest } from "./test-utils"; customTest.skip("case", () => {});`},
			{Code: `test.for([1]).skip("case", () => {}); describe.each([1]).skip("suite", () => {});`},
		},
		[]rule_tester.InvalidTestCase{
			{
				Code: `test.skip("case", () => {});`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "skippedTest", Line: 1, Column: 1},
				},
			},
			{
				Code: `it.skip("case", () => {});`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "skippedTest", Line: 1, Column: 1},
				},
			},
			{
				Code: `describe.skip("suite", () => {});`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "skippedTest", Line: 1, Column: 1},
				},
			},
			{
				Code: `test["skip"]("case", () => {}); it[` + "`skip`" + `]("case", () => {});`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "skippedTest", Line: 1, Column: 1},
					{MessageId: "skippedTest", Line: 1, Column: 33},
				},
			},
			{
				Code: `test.concurrent.skip("case", () => {}); test.skip.sequential("case", () => {}); test.fails.skip("case", () => {});`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "skippedTest", Line: 1, Column: 1},
					{MessageId: "skippedTest", Line: 1, Column: 41},
					{MessageId: "skippedTest", Line: 1, Column: 81},
				},
			},
			{
				Code: `test.skip.only("case", () => {}); test.only.skip("case", () => {});`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "skippedTest", Line: 1, Column: 1},
					{MessageId: "skippedTest", Line: 1, Column: 35},
				},
			},
			{
				Code: `describe.concurrent.skip("suite", () => {}); describe.skip.sequential("suite", () => {});`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "skippedTest", Line: 1, Column: 1},
					{MessageId: "skippedTest", Line: 1, Column: 46},
				},
			},
			{
				Code: `test.skip.each([1])("case", () => {}); test.skip.for([1])("case", () => {});`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "skippedTest", Line: 1, Column: 1},
					{MessageId: "skippedTest", Line: 1, Column: 40},
				},
			},
			{
				Code: `describe.skip.each([1])("suite", () => {}); describe.skip.for([1])("suite", () => {});`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "skippedTest", Line: 1, Column: 1},
					{MessageId: "skippedTest", Line: 1, Column: 45},
				},
			},
			{
				Code: "test.skip.each`value\n${1}`(\"case\", () => {});",
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "skippedTest", Line: 1, Column: 1},
				},
			},
			{
				Code: `import { test as rstestTest } from "@rstest/core";
rstestTest.skip("case", () => {});`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "skippedTest", Line: 2, Column: 1},
				},
			},
			{
				Code: `const { test: rstestTest } = require("@rstest/core");
rstestTest.skip("case", () => {});`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "skippedTest", Line: 2, Column: 1},
				},
			},
			{
				Code: `import * as rstest from "@rstest/core";
rstest.test.skip("case", () => {});
rstest.describe.skip("suite", () => {});`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "skippedTest", Line: 2, Column: 1},
					{MessageId: "skippedTest", Line: 3, Column: 1},
				},
			},
			{
				Code: `const rstest = require("@rstest/core");
rstest.test.skip("case", () => {});`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "skippedTest", Line: 2, Column: 1},
				},
			},
			{
				Code: `const skipped = test.skip;
skipped("case", () => {});`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "skippedTest", Line: 2, Column: 1},
				},
			},
			{
				Code: `const base = test;
const skipped = base.skip;
skipped.each([1])("case", () => {});`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "skippedTest", Line: 3, Column: 1},
				},
			},
			{
				Code: `import * as rstest from "@rstest/core";
const skippedSuite = rstest.describe.skip;
skippedSuite("suite", () => {});`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "skippedTest", Line: 3, Column: 1},
				},
			},
			{
				Code: `test.extend({}).skip("case", () => {});`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "skippedTest", Line: 1, Column: 1},
				},
			},
			{
				Code: `const appTest = test.extend({});
const adminTest = appTest.extend({});
adminTest.concurrent.skip("case", () => {});`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "skippedTest", Line: 3, Column: 1},
				},
			},
			{
				Code: `test("case"); it("case"); test.concurrent("case"); test.fails("case");`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "missingFunction", Line: 1, Column: 1},
					{MessageId: "missingFunction", Line: 1, Column: 15},
					{MessageId: "missingFunction", Line: 1, Column: 27},
					{MessageId: "missingFunction", Line: 1, Column: 52},
				},
			},
			{
				Code: `test.each([1])("case"); test.for([1])("case");`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "missingFunction", Line: 1, Column: 1},
					{MessageId: "missingFunction", Line: 1, Column: 25},
				},
			},
			{
				Code: `test("case", { timeout: 100 }); it("case", {}); test.concurrent("case", { retry: 2 });`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "missingFunction", Line: 1, Column: 1},
					{MessageId: "missingFunction", Line: 1, Column: 33},
					{MessageId: "missingFunction", Line: 1, Column: 49},
				},
			},
			{
				Code: `test.each([1])("case", { timeout: 100 });`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "missingFunction", Line: 1, Column: 1},
				},
			},
			{
				Code: `test.skip("case", { timeout: 100 });`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "skippedTest", Line: 1, Column: 1},
					{MessageId: "missingFunction", Line: 1, Column: 1},
				},
			},
			{
				Code: `const customTest = test.extend({});
customTest("case");`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "missingFunction", Line: 2, Column: 1},
				},
			},
			{
				Code: `test.skip("case");`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "skippedTest", Line: 1, Column: 1},
					{MessageId: "missingFunction", Line: 1, Column: 1},
				},
			},
			{
				Code: `test.todo.skip("case");`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "skippedTest", Line: 1, Column: 1},
				},
			},
		},
	)
}

File: packages/rslint-test-tools/tests/rstest/rules/no-disabled-tests.test.ts
import { RuleTester } from '../rule-tester';

const ruleTester = new RuleTester();

ruleTester.run('no-disabled-tests', {} as never, {
  valid: [
    {
      code: 'test("case", () => {}); it("case", () => {}); describe("suite", () => {});',
    },
    { code: 'test.only("case", () => {}); test.fails("case", () => {});' },
    {
      code: 'describe.only("suite", () => {}); describe.concurrent.sequential("suite", () => {});',
    },
    { code: 'test.todo("case"); it.todo("case"); describe.todo("suite");' },
    { code: 'test.todo.each([1])("case"); test.todo.for([1])("case");' },
    { code: 'const todoTest = test.todo; todoTest("case");' },
    {
      code: 'const condition = true; test.skipIf(condition)("case", () => {}); test.runIf(condition)("case", () => {});',
    },
    {
      code: 'const condition = true; describe.skipIf(condition)("suite", () => {}); describe.runIf(condition)("suite", () => {});',
    },
    // `skip` on the test context is a runtime decision, not a disabled test.
    { code: 'test("case", context => { context.skip(); });' },
    { code: 'describe("empty suite"); describe.concurrent("empty suite");' },
    { code: 'test("case", { timeout: 100 }, () => {});' },
    // an identifier is not necessarily an options object
    { code: 'declare const options: any; test("case", options);' },
    // jest-only aliases are not rstest APIs
    {
      code: 'declare const fit: any, xit: any, xtest: any, fdescribe: any, xdescribe: any; fit("case", () => {}); xit("case", () => {}); xtest("case", () => {}); fdescribe("suite", () => {}); xdescribe("suite", () => {});',
    },
    { code: 'declare const pending: any; pending();' },
    // dynamic member access cannot be resolved to `skip`
    {
      code: 'declare const suffix: string; test[`sk${suffix}`]("case", () => {});',
    },
    // not the rstest `test`
    { code: 'import { test } from "node:test"; test.skip("case", () => {});' },
    {
      code: 'import { test, describe } from "vitest"; test.skip("case", () => {}); describe.skip("suite", () => {});',
    },
    {
      code: 'declare function createRunner(): any; const test = createRunner(); test.skip("case", () => {});',
    },
    // `.skip` after `.each`/`.for` is not a valid rstest chain
    {
      code: 'test.for([1]).skip("case", () => {}); describe.each([1]).skip("suite", () => {});',
    },
  ],
  invalid: [
    {
      code: 'test.skip("case", () => {});',
      errors: [
        {
          line: 1,
          column: 1,
          messageId: 'skippedTest',
          message: 'Tests should not be skipped',
        },
      ],
    },
    {
      code: 'it.skip("case", () => {});',
      errors: [{ line: 1, column: 1, messageId: 'skippedTest' }],
    },
    {
      code: 'describe.skip("suite", () => {});',
      errors: [{ line: 1, column: 1, messageId: 'skippedTest' }],
    },
    {
      code: 'test["skip"]("case", () => {}); it[`skip`]("case", () => {});',
      errors: [
        { line: 1, column: 1, messageId: 'skippedTest' },
        { line: 1, column: 33, messageId: 'skippedTest' },
      ],
    },
    {
      code: 'test.concurrent.skip("case", () => {}); test.skip.sequential("case", () => {});',
      errors: [
        { line: 1, column: 1, messageId: 'skippedTest' },
        { line: 1, column: 41, messageId: 'skippedTest' },
      ],
    },
    {
      code: 'test.skip.each([1])("case", () => {}); test.skip.for([1])("case", () => {});',
      errors: [
        { line: 1, column: 1, messageId: 'skippedTest' },
        { line: 1, column: 40, messageId: 'skippedTest' },
      ],
    },
    {
      code: 'describe.skip.each([1])("suite", () => {});',
      errors: [{ line: 1, column: 1, messageId: 'skippedTest' }],
    },
    {
      code: [
        'import { test as rstestTest } from "@rstest/core";',
        'rstestTest.skip("case", () => {});',
      ].join('\n'),
      errors: [{ line: 2, column: 1, messageId: 'skippedTest' }],
    },
    {
      code: [
        'import * as rstest from "@rstest/core";',
        'rstest.test.skip("case", () => {});',
        'rstest.describe.skip("suite", () => {});',
      ].join('\n'),
      errors: [
        { line: 2, column: 1, messageId: 'skippedTest' },
        { line: 3, column: 1, messageId: 'skippedTest' },
      ],
    },
    {
      code: ['const skipped = test.skip;', 'skipped("case", () => {});'].join(
        '\n',
      ),
      errors: [{ line: 2, column: 1, messageId: 'skippedTest' }],
    },
    {
      code: 'test.extend({}).skip("case", () => {});',
      errors: [{ line: 1, column: 1, messageId: 'skippedTest' }],
    },
    {
      code: 'test("case"); it("case"); test.concurrent("case");',
      errors: [
        {
          line: 1,
          column: 1,
          messageId: 'missingFunction',
          message: 'Test is missing function argument',
        },
        { line: 1, column: 15, messageId: 'missingFunction' },
        { line: 1, column: 27, messageId: 'missingFunction' },
      ],
    },
    {
      code: 'test.each([1])("case"); test.for([1])("case");',
      errors: [
        { line: 1, column: 1, messageId: 'missingFunction' },
        { line: 1, column: 25, messageId: 'missingFunction' },
      ],
    },
    // `(description, options)` registers a test without a body
    {
      code: 'test("case", { timeout: 100 }); it("case", {});',
      errors: [
        { line: 1, column: 1, messageId: 'missingFunction' },
        { line: 1, column: 33, messageId: 'missingFunction' },
      ],
    },
    // both a skipped test and a missing implementation
    {
      code: 'test.skip("case");',
      errors: [
        { line: 1, column: 1, messageId: 'skippedTest' },
        { line: 1, column: 1, messageId: 'missingFunction' },
      ],
    },
    // `.todo` suppresses only the missing-function report
    {
      code: 'test.todo.skip("case");',
      errors: [{ line: 1, column: 1, messageId: 'skippedTest' }],
    },
  ],
});
